test/EndToEnd/tests/PackageNameSpaceTests.ps1

```diff
@@ -0,0 +1,246 @@
+function Test-PackageNameSpaceRestore
+{
+    param($context)
+
+    # Arrange
+    $repoDirectory = Join-Path $OutputPath "CustomPackages"
+    $opensourceRepo = Join-Path $repoDirectory "opensourceRepo"
+    $privateRepo = Join-Path $repoDirectory "privateRepo"
+    $nugetConfigPath = Join-Path $OutputPath 'nuget.config'
+
+    $settingFileContent =@"
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+    <packageSources>
+    <clear />
+    <add key="PublicRepository" value="{0}" />
+    <add key="PrivateRepository" value="{1}" />
+    </packageSources>
+    <packageNamespaces>
+        <packageSource key="PublicRepository"> 
+            <namespace id="Contoso.Opensource.*" />
+        </packageSource>
+        <packageSource key="PrivateRepository">
+            <namespace id="Contoso.MVC.*" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>
+"@    
+    try {
+
+        # We have to create config file before creating solution, otherwise it's not effective for new solutions.
+        $settingFileContent -f $opensourceRepo,$privateRepo | Out-File -Encoding "UTF8" $nugetConfigPath
+    
+        $p = New-ConsoleApplication
+    
+        $projectDirectoryPath = $p.Properties.Item("FullPath").Value
+        $packagesConfigPath = Join-Path $projectDirectoryPath 'packages.config'
+        $projectDirectoryPath = $p.Properties.Item("FullPath").Value
+        $solutionDirectory = Split-Path -Path $projectDirectoryPath -Parent   
+        # Write a file to disk, but do not add it to project
+        '<packages>
+            <package id="Contoso.MVC.ASP" version="1.0.0" targetFramework="net461" />
+            <package id="Contoso.Opensource.Buffers" version="2.0.0" targetFramework="net461" />
+            <package id="Foo" version="1.0.0" targetFramework="net461" />
+    </packages>' | out-file $packagesConfigPath
+    
+    
+        CreateCustomTestPackage "Contoso.MVC.ASP" "1.0.0" $privateRepo
```

Comment:
Would it be easier to use the already preexisting packages that come through the package test data?

> $root\packages\nuget.client.endtoend.testdata\1.0.0\content\Packages

Reply:
Personally, I'm not too worried given that it's only 1 test though.

Reply:
Changed to pre-existing package.

Reply:
Of course I can remove this one. Please let me know.